plasmapy/diagnostics/thomson.py

```diff
@@ -204,114 +444,422 @@ def spectral_density(
+            * Te.unit
+        )
+
     # Make sure the sizes of efract, electron_vel, and Te all match
     if (electron_vel.shape[0] != efract.size) or (Te.size != efract.size):
         raise ValueError(
             f"Inconsistent number of electron populations in efract ({efract.size}), "
             f"Te ({Te.size}), or electron velocity ({electron_vel.shape[0]})."
         )
 
-    # Ensure unit vectors are normalized
+    # Create arrays of ion Z and mass from particles given
+    ion_z = np.zeros(len(ion_species))
+    ion_mass = np.zeros(len(ion_species)) * u.kg
+    for i, particle in enumerate(ion_species):
+        ion_z[i] = particle.charge_number
+        ion_mass[i] = particle_mass(particle)
+
     probe_vec = probe_vec / np.linalg.norm(probe_vec)
     scatter_vec = scatter_vec / np.linalg.norm(scatter_vec)
 
-    # Define some constants
-    C = const.c.si  # speed of light
+    # Apply the insturment function
+    if instr_func is not None and callable(instr_func):
+
+        # Create an array of wavelengths of the same size as wavelengths
+        # but centered on zero
+        wspan = (np.max(wavelengths) - np.min(wavelengths)) / 2
+        eval_w = np.linspace(-wspan, wspan, num=wavelengths.size)
+        instr_func_arr = instr_func(eval_w)
+
+        if type(instr_func_arr) != np.ndarray:
+            raise ValueError(
+                "instr_func must be a function that returns a "
+                "np.ndarray, but the provided function returns "
+                f" a {type(instr_func_arr)}"
+            )
+
+        if wavelengths.shape != instr_func_arr.shape:
+            raise ValueError(
+                "The shape of the array returned from the "
+                f"instr_func ({instr_func_arr.shape}) "
+                "does not match the shape of the wavelengths "
+                f"array ({wavelengths.shape})."
+            )
+
+        instr_func_arr *= 1 / np.sum(instr_func_arr)
+    else:
+        instr_func_arr = None
+
+    alpha, Skw = spectral_density_lite(
+        wavelengths.to(u.m).value,
+        probe_wavelength.to(u.m).value,
+        n.to(u.m ** -3).value,
+        Te.to(u.K).value,
+        Ti.to(u.K).value,
+        efract=efract,
+        ifract=ifract,
+        ion_z=ion_z,
+        ion_mass=ion_mass.to(u.kg).value,
+        ion_vel=ion_vel.to(u.m / u.s).value,
+        electron_vel=electron_vel.to(u.m / u.s).value,
+        probe_vec=probe_vec,
+        scatter_vec=scatter_vec,
+        instr_func_arr=instr_func_arr,
+    )
+
+    return alpha, Skw * u.s / u.rad
+
+
+# ***************************************************************************
+# These functions are necessary to interface scalar Parameter objects with
+# the array inputs of spectral_density
+# ***************************************************************************
+
+
+def _count_populations_in_params(params, prefix):
+    """
+    Counts the number of entries matching the pattern prefix_i in a
+    list of keys
```

Comment:
```suggestion
    Counts the number of items in the ``params`` `dict` with a key
    that starts with the string defined by ``prefix``.
```

Reworded the docstring so the argument descriptions are contained in the one line function description.  **Note,** I know `params` will likely be an instance of `lmfit.parameter.Parameters`, but `Parameters` is a subclass of `dict` and this functionality works with dictionaries so I kept the description more general.